src/imitation/algorithms/preference_comparisons.py

```diff
@@ -919,16 +963,43 @@ def _make_data_loader(self, dataset: PreferenceDataset) -> data_th.DataLoader:
 
     def _train(self, dataset: PreferenceDataset, epoch_multiplier: float = 1.0) -> None:
         """Trains for `epoch_multiplier * self.epochs` epochs over `dataset`."""
-        dataloader = self._make_data_loader(dataset)
+        if self.regularizer:
+            val_length = int(len(dataset) * self.val_split)
+            train_length = len(dataset) - val_length
+            train_dataset, val_dataset = data_th.random_split(
+                dataset,
+                lengths=[train_length, val_length],
+                generator=th.Generator().manual_seed(self.seed),
+            )
+            dataloader = self._make_data_loader(train_dataset)
+            val_dataloader = self._make_data_loader(val_dataset)
+        else:
+            dataloader = self._make_data_loader(dataset)
+            val_dataloader = None
+
         epochs = round(self.epochs * epoch_multiplier)
 
         for _ in tqdm(range(epochs), desc="Training reward model"):
+            train_loss = 0.0
+            val_loss = 0.0
             for fragment_pairs, preferences in dataloader:
                 self.optim.zero_grad()
                 loss = self._training_inner_loop(fragment_pairs, preferences)
-                loss.backward()
+                train_loss += loss.item()
+                if self.regularizer:
+                    self.regularizer.regularize(loss)
                 self.optim.step()
 
+            if self.regularizer:
+                # TODO(juan) my type checker complains because it cannot infer the type
+                #  of val_dataloader by checking the logic. Is there a better way
+                #  to do this, or should we leave the cast?
+                val_dataloader = cast(data_th.DataLoader, val_dataloader)
```

Comment:
Can it follow it if you `assert val_dataloader is not None` (should work if it's inferred type as `Optional[data_th.DataLoader]`)?

That's a bit less aggressive than a cast in that if you earlier on do something like `val_dataloader = "foo"` it'll still fail type checking.

Reply:
I was not a big fan of using assert in that pattern. However, I had not really thought about cast vs assert much. I'm now of the opinion that asserting is 'less evil' than asserting whenever assert can be used, i.e. when this is literals or class instances. It doesn't work with type vars / generics / protocols, where we should use cast.

Reply:
I'm surprised type checkers cannot infer that this is not none, given what other impressive things they can do. This is a general pattern that they are consistently bad at.

Reply:
In fact Pylance continues to complain after I add the assert, but that's just because Pylance is not very good.